File: Java/Iterator/BookCollection/src/com/ericgtkb/BookCollection.java
package com.ericgtkb;

import java.util.ArrayList;
import java.util.List;

public interface BookCollection {
    CollectionIterator createIterator();
}

class ScienceCollection implements BookCollection {
    private Book[] books;

    public ScienceCollection() {
        books = new Book[3];
        books[0] = new Book("Physics");
        books[1] = new Book("Math");
        books[2] = new Book("Chemistry");
    }

    @Override
    public CollectionIterator createIterator() {
        return new ScienceCollectionIterator(books);
    }
}

class MusicCollection implements BookCollection {
    private List<Book> bookList;

    public MusicCollection() {
        bookList = new ArrayList<>();
        bookList.add(new Book("Scales and Modes"));
        bookList.add(new Book("Jazz History"));
        bookList.add(new Book("Contemporary Harmony"));
        bookList.add(new Book("Music Theory"));
    }

    @Override
    public CollectionIterator createIterator() {
        return new MusicCollectionIterator(bookList);
    }
}

File: C++/Iterator/BookCollection/book_collection.cc
#include "book_collection.h"

ScienceCollection::ScienceCollection() : _books {Book("Physics"), Book("Math"), Book("Chemistry")},
                                         _size {sizeof(_books) / sizeof(_books[0])} {}

std::unique_ptr<CollectionIterator> ScienceCollection::create_iterator() const {
    return std::make_unique<ScienceCollectionIterator>(_books, _size);
}

MusicCollection::MusicCollection() : _books {Book("Scales and Modes"),
                                             Book("Jazz History"),
                                             Book("Contemporary Harmony"),
                                             Book("Music Theory")} {}

std::unique_ptr<CollectionIterator> MusicCollection::create_iterator() const {
    return std::make_unique<MusicCollectionIterator>(_books);
}
